openedx/core/djangoapps/content/learning_sequences/api/processors/base.py

```diff
@@ -0,0 +1,80 @@
+"""
+This module defines the base OutlineProcessor class that is the primary method
+of adding new logic that manipulates the Course Outline for a given student.
+"""
+import logging
+from datetime import datetime
+
+from django.contrib.auth import get_user_model
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+from ..data import ScheduleData, ScheduleItemData
+
+User = get_user_model()
+log = logging.getLogger(__name__)
+
+
+class OutlineProcessor:
+    """
+    Base class for manipulating the Course Outline.
+
+    You can inherit from this class and extend any of its four main methods:
+    __init__, load_data, inaccessible_sequences, usage_keys_to_remove.
+
+    An OutlineProcessor is invoked synchronously during a request for the
+    CourseOutline. The steps are:
+        * __init__
+        * load_data
+        * inaccessible_sequences, usage_keys_to_remove (no ordering guarantee)
+
+    Also note that you should not assume any ordering relative to any other
+    OutlineProcessor. Once async support works its way fully into Django, we'll
+    likely even want to run these in parallel.
```

Comment:
Ah, so maybe not trim after every loop? ^

Reply:
Yeah, but the performance implications are only secondary to that I think. For me, the most important reason to keep things independent are predictability and debugging. I think that it'll be easier to determine the correctness of something that always operates on the entire outline, rather than something that operates on a subset that assumes certain rules have already been applied.

FWIW, on the performance front, I expect that the vast majority of the overhead is going to be in the `load_data()` methods, since they will probably involve database access to dredge up data. `remove()` is cheap. Set/dict operations on opaque keys are way more expensive than they have any right to be, but we're only dealing with hundreds of them at a time. There's some perf tuning work being done on that by the always-available team as well, which should help that.